torch/csrc/jit/tensorexpr/kernel.cpp

```diff
@@ -1713,6 +1713,59 @@ Tensor* TensorExprKernel::computeValue(const torch::jit::Value* v) {
   }
 }
 
+// Return the (lower, upper) loop bounds if they are constants, else nullopt.
+c10::optional<std::pair<int64_t, int64_t>> loopBounds(const For* loop) {
+  auto start = IRSimplifier::simplify(loop->start());
+  auto stop = IRSimplifier::simplify(loop->stop());
+  if (!start->isConstant() || !stop->isConstant()) {
+    return c10::nullopt;
+  }
+  return c10::make_optional(
+      std::make_pair(immediateAs<int64_t>(start), immediateAs<int64_t>(stop)));
+}
+
+// True if all the loops in this vector have equal bounds.
+bool loopBoundsAllEqual(const std::vector<For*>& loops) {
+  auto bounds = loopBounds(loops[0]);
+  if (!bounds) {
+    return false;
+  }
+  for (auto const& loop : loops) {
+    auto next = loopBounds(loop);
+    if (!next) {
+      return false;
+    }
+    if (bounds->first != next->first || bounds->second != next->second) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// Recursively fuse all the loops with matching bounds in `st`.  Stops fusing
+// at any level containing non-loops or non-matching bounds.  The restriction
+// on matching bounds exists to avoid inserting conditionals on the loop
+// indices where none would be needed, which would significantly complicate
+// vectorization.
+void fuseAllLoops(Stmt* st) {
+  if (auto block = dynamic_cast<tensorexpr::Block*>(st)) {
+    std::vector<For*> loopsToFuse;
+    for (auto stmt : *block) {
+      auto loop = dynamic_cast<For*>(stmt);
+      if (!loop) {
+        // Block contains something that's not a loop.  Quit.
+        return;
+      }
```

Comment:
This is going to stop if the block contains a non-loop statement anywhere. We could try to fuse loops that are together even though there are other statements before / after these loops in the block (not sure if this could happen in our IR currently, but still.. ). 

Also, if there are 4 loops and 2 of them have same bounds, we could fuse them together at least.. 

If these are not necessary for your benchmark, just leave a TODO and I will take care of them.

Reply:
Yeah this is a good idea, and might improve the benchmark even further.